### PR TITLE
[BottomSheet] Add property so clients can set a custom height for bottom sheet

### PR DESCRIPTION
**Pull request by codeman7**
Previously clients could only set a custom height for bottom sheet by setting the `preferredContentSize` this adds a property so clients can set a specific height and not touch `preferredContentSize`.

Related to #4945

### CODE REVIEW

**Review by jverkoey — COMMENTED** [review]
We may need a test that checks what happens when `preferredSheetHeight` is set to a positive value and then back to zero.

File: components/BottomSheet/src/MDCBottomSheetPresentationController.h [hunk 1]
```diff
@@ -64,6 +64,15 @@
  */
 @property(nonatomic, assign) BOOL dismissOnBackgroundTap;
 
+/**
+ This is used to set a custom height on the sheet view.
+
+ @note This will override setting the preferredContentSize if a positive value is passed.
+ @note If a non-positive value is passed then the sheet will open up to either half the screen
+ height or the size of the contentViewController whatever value is smaller.
```
Comment (jverkoey) [inline components/BottomSheet/src/MDCBottomSheetPresentationController.h]: tip: I find documentation like this can be a bit easier to parse if written as a logical statement, like so:

> If this, then this. Otherwise, this.

So in this case:

> If a positive value is passed, overrides preferredContentSize. Otherwise, the sheet will open up to...

File: components/BottomSheet/src/MDCBottomSheetPresentationController.h [hunk 2]
```diff
@@ -64,6 +64,15 @@
  */
 @property(nonatomic, assign) BOOL dismissOnBackgroundTap;
 
+/**
+ This is used to set a custom height on the sheet view.
+
+ @note This will override setting the preferredContentSize if a positive value is passed.
```
Comment (jverkoey) [inline components/BottomSheet/src/MDCBottomSheetPresentationController.h]: What does override mean? Please add the answer as documentation.

File: components/BottomSheet/src/MDCBottomSheetPresentationController.m [hunk 3]
```diff
@@ -179,9 +179,11 @@ - (void)viewWillTransitionToSize:(CGSize)size
 
  @param preferredSheetHeight If positive, the new value for @sheetView.preferredSheetHeight.
  */
-- (void)setPreferredSheetHeight:(CGFloat)preferredSheetHeight {
+- (void)updatePreferredSheetHeight:(CGFloat)preferredSheetHeight {
```
Comment (jverkoey) [inline components/BottomSheet/src/MDCBottomSheetPresentationController.m]: I think this method doesn't need an argument because you're able to pull the information from either `self.presentedViewController.preferredContentSize.height` or `self.preferredSheetHeight` depending on the value of `self.preferredSheetHeight`. Removing this argument may simplify both this method and the call sites as a result.

File: components/BottomSheet/src/MDCBottomSheetPresentationController.m [hunk 4]
```diff
@@ -244,6 +246,13 @@ - (UIAccessibilityTraits)scrimAccessibilityTraits {
   return _scrimAccessibilityTraits;
 }
 
+- (void)setPreferredSheetHeight:(CGFloat)preferredSheetHeight {
+  _preferredSheetHeight = preferredSheetHeight;
+  if (preferredSheetHeight > 0.f) {
+    [self updatePreferredSheetHeight:preferredSheetHeight];
+  }
```
Comment (jverkoey) [inline components/BottomSheet/src/MDCBottomSheetPresentationController.m]: You can remove the if statement here if `updatePreferredSheetHeight` is changed to a `updatePreferredSheetHeight`-style method.

File: components/BottomSheet/src/MDCBottomSheetPresentationController.m (hunk 3 above)
Comment (jverkoey) [reply components/BottomSheet/src/MDCBottomSheetPresentationController.m]: The method name in this case could likely be renamed to `updatePreferredSheetHeight` or something similar to indicate that the value has already changed and we now want to update the sheetView's preferredSheetHeight property accordingly.

**Review by jverkoey — DISMISSED** [review]

File: components/BottomSheet/src/MDCBottomSheetPresentationController.h [hunk 5]
```diff
@@ -64,6 +64,15 @@
...
+ @note If a non-positive value is passed then the sheet will open up to either half the screen
+ height or the size of the contentViewController whatever value is smaller.
+ */
```
Comment (jverkoey) [inline components/BottomSheet/src/MDCBottomSheetPresentationController.h]: Looking at the implementation of MDCSheetContainerView.m, it looks like `truncatedPreferredSheetHeight` always considers the contentViewController's height when determining the maximum height. This implies that `preferredSheetHeight` will also be capped to the content's height, which the documentation doesn't yet indicate. We may want to add a test and documentation indicating this.

E.g. pseudo-code:

```
preferredSheetHeight = 1000
contentViewController.scrollView.contentSize.height = 500

assertEqual(sheet.height, 500) // Docs presently imply 1000
```

**Review by jverkoey — COMMENTED** [review]
I'm afraid I'm still not certain I understand when someone would set `preferredSheetHeight` vs `preferredContentSize.height`. As best I can tell from the code these two properties have equivalent results.

Can you share some additional context that might help me understand the need for this new property?

File: components/BottomSheet/src/MDCBottomSheetPresentationController.h (hunk 1 above)
Comment (codeman7) [reply components/BottomSheet/src/MDCBottomSheetPresentationController.h]: Done

File: components/BottomSheet/src/MDCBottomSheetPresentationController.h (hunk 2 above)
Comment (codeman7) [reply components/BottomSheet/src/MDCBottomSheetPresentationController.h]: I think I made this more clear.

File: components/BottomSheet/src/MDCBottomSheetPresentationController.m (hunk 3 above)
Comment (codeman7) [reply components/BottomSheet/src/MDCBottomSheetPresentationController.m]: I believe #5125 updated this method to make it more testable, I can revert that change if you'd like.

File: components/BottomSheet/src/MDCBottomSheetPresentationController.m (hunk 4 above)
Comment (codeman7) [reply components/BottomSheet/src/MDCBottomSheetPresentationController.m]: Pending your feedback on reverting #5125.

**Comment by codeman7** [comment]
Setting the `preferredContentSize` also effects the frame of the `UIViewController.view` so when you rotate or change `view.frame` it isn't correct, it will be the value of `preferredContentSize`. This makes it so that the `view.frame` will always be correct.

**Comment by jverkoey** [comment]
Ah! Interesting. Who sets view.frame in reaction to preferredContentSize? Is that us?

**Comment by codeman7** [comment]
In order to close #4945 I would either have to set `preferredContentSize` or this new property.

File: components/BottomSheet/src/MDCBottomSheetPresentationController.m (hunk 3 above)
Comment (jverkoey) [reply components/BottomSheet/src/MDCBottomSheetPresentationController.m]: Ah! I see.

I think we don't need to revert that change - you're increasing the surface area of the APIs here to include both a setter and a "did change" method that commits the results to the sheetView.

Prior to this PR, the tests relied on the method previously named `setPreferredSheetHeight`. The tests were using this method to propagate changes to the sheetView.

After this PR, `setPreferredSheetHeight` has a different meaning and you're introducing a new method, `updatePreferredSheetHeight` that models the old meaning of `setPreferredSheetHeight`.

You can still test as before, but your existing tests will now call the `updatePreferredSheetHeight` API to propagate changes to the sheetView instead of `setPreferredSheetHeight`, except in cases where you're testing the new property's behavior specifically.

Comment (jverkoey) [reply components/BottomSheet/src/MDCBottomSheetPresentationController.m]: Tying this back to my original suggestion, if you remove `updatePreferredSheetHeight`'s argument then your tests will still be able to call this method, but prior to doing so they must set the relevant properties that they are testing (either `preferredSheetHeight` or `preferredContentSize.height`).

Comment (jverkoey) [reply components/BottomSheet/src/MDCBottomSheetPresentationController.m]: Spoke over gvc:

1. Partial revert of #5125 so that we still have an `updatePreferredSheetHeight` method that doesn't accept arguments.
2. `updatePreferredSheetHeight` will be exposed to tests so that we can "flush" the calculated sheet height to sheetView once we've configured either `preferredSheetHeight` or `preferredContentSize.height`.

Comment (jverkoey) [reply components/BottomSheet/src/MDCBottomSheetPresentationController.m]: Also discussed the possibility of making `updatePreferredSheetHeight` more testable by making it stateless, but that doing so can be a follow-up exploration:

```objc
+ (CGFloat)preferredSheetHeightForViewController:(UIViewController *)viewController
                                       withFrame:(CGRect)frame
                            preferredSheetHeight:(CGFloat)preferredSheetHeight {
  CGFloat calculatedPreferredSheetHeight =
      (preferredSheetHeight > 0.f) ? preferredSheetHeight : viewController.preferredContentSize.height;
  if (MDCCGFloatEqual(calculatedPreferredSheetHeight, 0)) {
    calculatedPreferredSheetHeight = MDCRound(CGRectGetHeight(frame) / 2);
  }
  return calculatedPreferredSheetHeight;
}
```

File: components/BottomSheet/tests/unit/MDCBottomSheetPresentationControllerPreferredSheetHeightTests.m [hunk 6]
```diff
@@ -107,10 +107,12 @@ - (void)testSetPreferredSheetHeightZeroWhenSheetViewHasUnstandardizedFrame {
 - (void)testSetPreferredSheetHeightPositiveValue {
```
Comment (romoore) [inline components/BottomSheet/tests/unit/MDCBottomSheetPresentationControllerPreferredSheetHeightTests.m]: Super-nit, please fix up the test method names. They are no longer testing `setPreferredSheetHeight`.

**Review by jverkoey — APPROVED** [review]
Awesome! I love how clear this is now.
